### PR TITLE
[TextGeneration] Samling arguments for generation

### PR DESCRIPTION
**Pull request by horheynm**
1. Add sampling into the text-generation pipeline
2. Refactor token generation flow

I also need https://github.com/neuralmagic/deepsparse/pull/1234 to land to check on failures caused by this pr if any

### CODE REVIEW

File: src/deepsparse/transformers/token_generator.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        if self.sampling_temperature != 1.0:
+            logits /= self.sampling_temperature
+            
+        if self.top_k:
+            logits = self.apply_top_k(logits)
+        if self.top_p: 
+            logits = self.apply_top_p(logits)
+            
+        # penalties here
+        if self.frequency_penalty != 0.0:
```
Comment (horheynm) [inline src/deepsparse/transformers/token_generator.py:49]: check if penalties can be applied before deterministic

File: src/deepsparse/transformers/engines/nl_decoder_engine.py [hunk 2]
```diff
@@ -202,10 +201,7 @@ def __call__(
         else:
             logits = out[0]
 
-        # select batch idx 0, batch is always 1
-        token = self.generate_token(logits=logits[0, -1, :])
-
-        return token, logits
+        return logits
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/engines/nl_decoder_engine.py:204]: Please update the return type of this function, given that it no longer outputs logits

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 3]
```diff
@@ -115,6 +116,21 @@ class Config:
         " tokens is generated). Set to `None` to ignore this parameter."
         " Default is `None`.",
     )
+    top_p: Optional[float] = Field(
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/pipelines/text_generation.py:119]: Let's be more verbose about those arguments. Put yourself in the shoes of a user, they need to have a good documentation of what those arguments do.

File: src/deepsparse/transformers/utils/token_generator.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+# limitations under the License.
+
+from typing import List
+
+import numpy
+
+from deepsparse.utils.data import numpy_softmax
+
+
+class TokenGenerator:
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/utils/token_generator.py:22]: We need docstrings for this function. What does it do, what is the shape of logits etc.

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 5]
```diff
@@ -450,20 +465,29 @@ def engine_forward(
...
         # names in this context
+
         with self.timer_manager.new_timer_context(total_inference=False) as timer:
             streamer = context.get("streamer")
 
             if not self.cache_support_enabled:
-                tokens, prompt_logits = self.multitoken_engine(engine_inputs)
-                return numpy.array([tokens]), prompt_logits
+                prompt_logits = self.multitoken_engine(engine_inputs)
+                token_generator = TokenGenerator(prompt_logits[0])
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/pipelines/text_generation.py:474]:
```suggestion
                token_generator = TokenGenerator(prompt_logits)
```
maybe let the `TokenGenerator` take care of indexing.

File: src/deepsparse/transformers/pipelines/token_classification.py [hunk 6]
```diff
@@ -525,7 +525,6 @@ def _get_tag(self, entity_name: str) -> Tuple[str, str]:
         return bi, tag
 
     def _group_entities(self, entities: List[dict]) -> List[dict]:
-
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/pipelines/token_classification.py:528]: redundant change

File: src/deepsparse/transformers/utils/token_generator.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        for token in tokens:
+            self.update_frequencies(token)
+
+    def update_frequencies(self, token: numpy.ndarray):
+        self.tokens.append(token)
+        self.token_frequencies[token] += 1
+
+    def generate(self, logits: numpy.ndarray) -> numpy.ndarray:
+        """
+        Samples a token from the logits using the sampling temperature.
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/utils/token_generator.py:53]: ...and much much more now. Let's document what new functionalities do.

File: src/deepsparse/transformers/utils/token_generator.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        if self.presence_penalty != 0.0:
+            logits = self.apply_presence_penalty(logits)
+
+        probs = self.numpy_softmax(logits)
+        token = numpy.random.choice(len(probs), p=probs)
+        self.update_frequencies(token)
+
+        return token
+
+    def apply_frequency_penalty(self, logits: numpy.ndarray):
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/utils/token_generator.py:83]: typing for returned value missing. In general, let's not forget about typing and proper docstrings.

**Review by dbogunowicz — CHANGES_REQUESTED** [review]
Designwise looks great. Let's clean up the code, make it nicer in terms of style. Also, before landing let's make sure that all the tests in the new LLM testing harness pass:

https://github.com/neuralmagic/deepsparse/pull/1216

This PR has potential to accidently break main pathways.

**Comment by dbogunowicz** [comment]
Also, PR description missing, let's make the PR title more descriptive.

File: src/deepsparse/transformers/utils/token_generator.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+        self.deterministic = deterministic
+        self.sampling_termperature = sampling_temperature
+        self.top_k = top_k
+        self.top_p = top_p
+        self.frequency_penalty = frequency_penalty
+        self.presence_penalty = presence_penalty
+        self.tokens = []
+        for token in tokens:
+            self.update_frequencies(token)
```
Comment (bfineran) [inline src/deepsparse/transformers/utils/token_generator.py:45]: probably not the most expensive, but you can rewrite using `numpy.unique(tokens, return_counts=True)` to vectorize

File: src/deepsparse/transformers/pipelines/text_generation.py (hunk 5 above)
Comment (horheynm) [reply src/deepsparse/transformers/pipelines/text_generation.py:474]: `TokenGenerator` should not take care of indexing. That means logits can be of any arbitrary shape. 

One change I can see is to just pass the length of the logit (51200) instead of the actual logit.

File: src/deepsparse/transformers/utils/token_generator.py (hunk 9 above)
Comment (horheynm) [reply src/deepsparse/transformers/utils/token_generator.py:45]:
```
tokens = [1,2,3,4,5]
>>> numpy.unique(tokens, return_counts=True)
(array([1, 2, 3, 4, 5]), array([1, 1, 1, 1, 1]))
```

and making it True will miss duplicates

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 10]
```diff
@@ -450,20 +465,29 @@ def engine_forward(
...
+                    token = token_generator.generate(prompt_logit)
+                return numpy.array([self.tokens]), prompt_logits
 
             else:
                 # run the prompt through
                 with timer.time(_TextGenerationTimings.PROMPT_PREFILL):
-                    tokens, prompt_logits = self.prompt_inference(engine_inputs)
+                    prompt_logits = self.prompt_inference(engine_inputs)
 
+            tokens = engine_inputs[0][engine_inputs[1].nonzero()].tolist()
```
[the remark this answers is not in the document]
Comment (horheynm) [reply src/deepsparse/transformers/pipelines/text_generation.py:503]: By end end of 
```
token_generator.generate(prompt_logits[-1][0, -1, :])
```
it generated the same tokens as main

[the remark this answers is not in the document]
Comment (dbogunowicz) [reply src/deepsparse/transformers/pipelines/text_generation.py:503]: Ok, this is probably harmless and will be caught in tests anyway.

File: src/deepsparse/transformers/utils/token_generator.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dbogunowicz) [reply src/deepsparse/transformers/utils/token_generator.py]: sounds good. if kvcachestorage lands after this PR I will make sure to take this into consideration

File: src/deepsparse/transformers/pipelines/text_generation.py (hunk 5 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/pipelines/text_generation.py]: It could take care of indexing quite nicely.
If logits.shape == 3 this always means (batch_size, sequence_len, vocab size)
elif logits.shape == 2 (sequence_len, vocab_size)
elif logit.shape == 1 (vocab_size)

File: src/deepsparse/transformers/pipelines/text_generation.py (hunk 3 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/pipelines/text_generation.py:119]: name is fine, its more about the description

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 11]
```diff
@@ -115,6 +116,29 @@ class Config:
...
+        description="Used for filtering generated tokens. Keep"
+        " top_k generated tokens. Default set to 0",
+    )
+    presence_penalty: Optional[float] = Field(
+        default=0.0,
+        description="Penalty applied for generating new token. Any existing"
+        " token results in the subtraction of its corresponding logit value."
+        " Default set to 0.0",
+    )
+    frquency_peanlty: Optional[float] = Field(
```
Comment (mgoin) [inline src/deepsparse/transformers/pipelines/text_generation.py:136]:
```suggestion
    frequency_penalty: Optional[float] = Field(
```

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 12]
```diff
@@ -414,7 +437,12 @@ def process_inputs(self, inputs: TextGenerationInput) -> List[numpy.ndarray]:
             include_prompt_logits=inputs.include_prompt_logits,
             callback=inputs.callback,
             stop=inputs.stop,
+            top_p=inputs.top_p,
+            top_k=inputs.top_k,
+            presence_penalty=inputs.presence_penalty,
+            frequency_penalty=inputs.presence_penalty,
```
Comment (mgoin) [inline src/deepsparse/transformers/pipelines/text_generation.py:443]:
```suggestion
            frequency_penalty=inputs.frequency_penalty,
```

File: src/deepsparse/transformers/utils/token_generator.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (horheynm) [reply src/deepsparse/transformers/utils/token_generator.py]: For next PR

**Review by bfineran — DISMISSED** [review]

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 13]
```diff
@@ -545,14 +596,13 @@ def prompt_inference(
...
-        new_token = None
         num_tokens_processed = 0
 
         if len(tokens) > self.prompt_sequence_length and self.enable_multitoken_prefill:
             self.multitoken_engine.reset_kv_cache()
             for engine_inputs in self.engine_inputs_for_prefill(tokens):
-                new_token, new_logits = self.multitoken_engine(engine_inputs)
-                num_tokens_processed += self.prompt_sequence_length
+                new_logits = self.multitoken_engine(engine_inputs)
+                num_tokens_processed += self.prompt_processing_sequence_length
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/pipelines/text_generation.py:605]:
```suggestion
                num_tokens_processed += self.prompt_sequence_length
```

**Review by dbogunowicz — APPROVED** [review]

**Comment by bfineran** [comment]
tests unrelated, merging

**Review by bfineran — APPROVED** [review]
